pkg/reconciler/namespace/namespace.go

```diff
@@ -137,26 +136,33 @@ func (c *Controller) updateSchedulableCondition(ctx context.Context, ns *corev1.
 // assignCluster attempts to schedule a namespace to a random cluster. If no cluster is
 // available, any existing cluster assignment will be cleared.
 func (c *Controller) assignCluster(ctx context.Context, ns *corev1.Namespace) error {
-	clusters, err := c.clusterLister.List(labels.Everything())
+	allClusters, err := c.clusterLister.List(labels.Everything())
 	if err != nil {
 		return err
 	}
 
-	oldClusterName := ns.Labels[clusterLabel]
+	var clusters []*v1alpha1.Cluster
+	for i := range allClusters {
+		// Only include Clusters that are in the same logical cluster as ns
+		if allClusters[i].ClusterName == ns.ClusterName {
+			clusters = append(clusters, allClusters[i])
+		}
+	}
 
 	// TODO: Filter out un-Ready clusters so a namespace doesn't
 	// get assigned to an un-Ready cluster.
 
 	newClusterName := ""
 	if len(clusters) > 0 {
 		// Select a cluster at random.
-		cluster := clusters[rand.Intn(len(clusters))]
+		cluster := allClusters[rand.Intn(len(allClusters))]
 		newClusterName = cluster.Name
 	}
 
+	oldClusterName := ns.Labels[clusterLabel]
 	if oldClusterName != newClusterName {
 		klog.Infof("Patching to update cluster assignment for namespace %q|%q: %q -> %q", ns.ClusterName, ns.Name, oldClusterName, newClusterName)
-		if _, err := c.namespaceClient.Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(newClusterName), metav1.PatchOptions{}); err != nil {
+		if _, err := c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(newClusterName), metav1.PatchOptions{}); err != nil {
```

Comment:
(No action required) Nothing to do with this PR, but I realize that I'm a bit confused at the reuse of the concept of 'cluster name' for both 'the namespace exists in a workspace identified by clustername' and 'the namespace is scheduled to a physical cluster identified by clustername'. Given that there is no way around using `meta.ClusterName`, maybe differentiation of the second case is suggested.

Reply:
"namespace is scheduled to a physical cluster identified by clustername" - clustername here should never be a meta.ClusterName, I don't think?